Get_Repo_Revision.py: about twice as fast, more work to do

permcody opened the pull request:
The get_repo_revision.py script has been speed up by roughly 2x just by changing the order of repository discovery (e.g. we shouldn't be trying SVN first anymore, or at all for that matter). 

More importantly, the second commit in this PR makes it so we don't generate revision files for modules, examples, tutorials, or moose/test. This drastically increases compilation speed when there is little or nothing to do.

permcody committed:
about twice as fast, more work to do

refs #10969

permcody committed:
Only generate revision files when necessary

closes #10696

friedmud reviewed (changes requested):
Needs `.md` documentation for this new Makefile parameter... including how to turn it on and an example of how to use what it generates.

permcody committed:
Modify the get_repo_revision.py file in two ways:

- Cache any shell command called more than once
- Modify the git repository files if nothing changes

permcody commented:
More notes for reviewer:
The changes to `get_repo_revision.py` are not as extensive as they appear. I made the script into a class so that I could cache a few of the shell calls that are made repeatedly so almost the entire script has been indented. I also change the odd spacing (e.g. "method( arg )" -> "method(arg)") throughout and of course had to add `self` in several places do to the promotion from functions to methods. The only real changes are at the bottom of the `writeRevision` method where we do extra stuff when the revision does **not** change.

Stork now defaults to **not** building a revision file.

brianmoose commented:
```
Traceback (most recent call last):
  File "/opt/civet/build_0/moose/framework/scripts/get_repo_revision.py", line 217, in <module>
    version_info.writeRevision(repo_location, app_name, header_file)
  File "/opt/civet/build_0/moose/framework/scripts/get_repo_revision.py", line 207, in writeRevision
    shellCommand("touch -r " + revision_header + " " + touch_file)
  File "/opt/civet/build_0/moose/framework/scripts/get_repo_revision.py", line 28, in shellCommand
    raise Exception()
Exception
```
Seems to show up on all the apps while building...
